### PR TITLE
customer agent interface

### PR DESCRIPTION
**Pull request by widyaaulia**
It currently only has one customer giving an order to multiple order-processors. The customer will "REQUEST" the amount of each product it wants to order. The order processor will "PROPOSE" with the price. The customer check the lowest price for each order product then send "CONFIRM" the cheapest bakery for respective product type.

### CODE REVIEW

**Comment by argenos** [comment]
Unfortunately this is not what the task for your team was, please read #27 carefully

**Comment by argenos** [comment]
> I am afraid I need to assure that we understand the task properly. Regarding the first task, does that mean that before the customer order anything, it has already received a list of product each bakery have? 
No, but this is precisely the job of the customer, to find out which products are produced by which bakery. 
 
>Then, after that it find the cheapest bakery and order from that? 
After that, you should split the order so you can order each product from the cheapest bakery. 

> I pull requested a whole as it needs some function and files to read, ex. the price list. So do we still need to include it or just the customeragent.java that we need to pull request?
You should be getting this from the other agents in the scenario. Your pull request needs to contain the bare minimum. 

> Regarding the customer, how should we make them? Should we create multiple customers at the start? Or should we create them while the program is running?
This is up to you. 

> Another thing, what does appropriate time mean?
If the order date is day 5, hour 12, your client should wait (using the TimeKeeper agent #45 ) only until day 5, hour 12, not before and not after.

File: README.md [hunk 1]
```diff
@@ -1,14 +1,20 @@
-# MAAS Project - <Team Name>
+[![Codacy Badge](https://api.codacy.com/project/badge/Grade/7e250960350b4dba8b4e031e3d2b2918)](https://app.codacy.com/app/ssmabidi/ws18-project-jsw?utm_source=github.com&utm_medium=referral&utm_content=HBRS-MAAS/ws18-project-jsw&utm_campaign=Badge_Grade_Dashboard)
```
Comment (argenos) [inline README.md]: This file should not be included in your pull request

File: build.gradle [hunk 2]
```diff
@@ -5,7 +5,7 @@ apply plugin: 'java'
 apply plugin: 'eclipse'
```
Comment (argenos) [inline build.gradle]: This file should not be included in your pull request

File: src/main/java/org/jsw/Start.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    	NameCollection nameCollection = new NameCollection();
+    	List<String> productType = nameCollection.getProductType();
+    	List<String> bakeryName = nameCollection.getBakeryName();
+    	
+    	//Create File of Price List for Each Bakery
+    	PriceList priceList = new PriceList();
+    	String fileName;
+    	for (String bakery : bakeryName) {
+    		try {
+    			fileName = "/home/widya/Gradle/ws18-project-jsw/src/main/resources/config/list/" + bakery + ".json";
```
Comment (argenos) [inline src/main/java/org/jsw/Start.java]: This should not be hard-coded.

File: src/main/java/org/jsw/Start.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			} catch (JSONException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+    	}
+    	
+    	List<String> agents = new Vector<>();
+    	agents.add("customer1:org.jsw.agents.CustomerAgent");
```
Comment (argenos) [inline src/main/java/org/jsw/Start.java]: Change this to org.maas.agents and move your file to that package.

File: src/main/java/org/jsw/Start.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    	
+    	for (String bakery : bakeryName) {
+    		agents.add(bakery + ":org.jsw.agents.OrderProcessingAgent");
```
Comment (argenos) [inline src/main/java/org/jsw/Start.java]: Same as above

File: src/main/java/org/jsw/agents/CustomerAgent.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
+package org.jsw.agents;
```
Comment (argenos) [inline src/main/java/org/jsw/agents/CustomerAgent.java]: move this to a `org.maas.agents` package

File: src/main/java/org/jsw/helpers/GenerateOrder.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class GenerateOrder{
+	private List<JSONObject> orders;
+	
+	/*
+	Orders are randomly generated. They are saved in JSON format and includes ID and ordered
```
Comment (argenos) [inline src/main/java/org/jsw/helpers/GenerateOrder.java]: The orders should be read from the provided config files.

File: src/main/java/org/jsw/helpers/ManageMessage.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
+package org.jsw.helpers;
```
Comment (argenos) [inline src/main/java/org/jsw/helpers/ManageMessage.java]: The naming of this file is confusing. Can you think of something else?

File: src/main/java/org/jsw/helpers/ManageMessage.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		"bakery_2": {"baguette": 10, ...},
+		...
+	} 
+	Process: pick the lowest offered price from all bakeries for each type of product
+	Expected Confirmation: 
+	{
+		"baguette": "bakery_1", "apple pie": "bakery_3", ...
+	}
+	*/
+	public static JSONObject findTheCheapest(JSONObject message, List<String> bakeryName, List<String> productTypes) throws JSONException {
```
Comment (argenos) [inline src/main/java/org/jsw/helpers/ManageMessage.java]: This should be done by asking other bakeries.

File: src/main/java/org/jsw/helpers/NameCollection.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
+package org.jsw.helpers;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class NameCollection {
```
Comment (argenos) [inline src/main/java/org/jsw/helpers/NameCollection.java]: These products should come from the scenario files.

File: src/main/java/org/jsw/helpers/PriceList.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+		int min = 5;
+		int max = 10;
+		
+		try {
+			order.put("Bakery Name", name);
+			
+			JSONObject product = new JSONObject();
+			
+			for(String type : productType ) {
+				price = ThreadLocalRandom.current().nextInt(min, max + 1);
```
Comment (argenos) [inline src/main/java/org/jsw/helpers/PriceList.java]: The scenario files have already specified the prices for the different bakeries, this is why you should get it from the agents directly!

**Comment by widyaaulia** [comment]
I see. However in my code, what I did is send the order to all sellers. Then we expect each sellers to return the price if they have it and zero if they don't have it. So when the customer receive a zero price means that the seller doesn't have it and then it will skip the minimum price checking for that bakery for the specific product type. Then the customer will send confirmation to order to the cheapest bakery for each product type.

May I know how the customer knows which bakery has which product? Should the customer read the json file listing all the product and the price each bakeries have and directly pick the cheapest ones? So the conversation between the customer and the seller will be just whether the seller is available or not?

**Comment by argenos** [comment]
> May I know how the customer knows which bakery has which product? 

You are already doing it! 

> I see. However in my code, what I did is send the order to all sellers. **Then we expect each sellers to return the price if they have it and zero if they don't have it.** So when the customer receive a zero price means that the seller doesn't have it and then it will skip the minimum price checking for that bakery for the specific product type. Then the customer will send confirmation to order to the cheapest bakery for each product type.


> Should the customer read the json file listing all the product and the price each bakeries have and directly pick the cheapest ones? So the conversation between the customer and the seller will be just whether the seller is available or not?

No. Information about what a bakery sells, and at which price, should be done through messages! Remember you will run only part of the scenario, other teams will have different agents and you won't be able to read from a file what they have.

**Comment by widyaaulia** [comment]
Indeed the customer knows what a bakery sells and which price through message. The findTheCheapest function is called by the customer and takes input from the message from all sellers combined together. The naming is I admit so confusing. I will change them.

While for the files in config, as there is only one example and I need to check whether the function I made is working, that is why I made some random lists. So I am guessing, the logic is right and now we just need to change, instead of making the order randomly, we need to use the files in config.

**Comment by argenos** [comment]
There’s been a config file for a couple of weeks now! But I agree, this should be the first thing you do after removing the unnecessary files

**Pull request closed without merging**